source/TestStep.cpp

```diff
@@ -0,0 +1,287 @@
+/* TestStep.cpp
+Copyright (c) 2019 by Peter van der Meer
+
+Endless Sky is free software: you can redistribute it and/or modify it under the
+terms of the GNU General Public License as published by the Free Software
+Foundation, either version 3 of the License, or (at your option) any later version.
+
+Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE.  See the GNU General Public License for more details.
+*/
+
+#include "ConditionSet.h"
+#include "DataNode.h"
+#include "Files.h"
+#include "GameData.h"
+#include "MainPanel.h"
+#include "Panel.h"
+#include "PlanetPanel.h"
+#include "PlayerInfo.h"
+#include "Ship.h"
+#include "TestData.h"
+#include "TestStep.h"
+#include "UI.h"
+#include <string>
+
+using namespace std;
+
+namespace {
+	bool SendFlightCommand(const Command &command, UI &gamePanels)
+	{
+		// We need to send the command to the top gamepanel. And it needs to be active.
+		if(gamePanels.IsEmpty() || gamePanels.Root() != gamePanels.Top())
+			return false;
+		
+		// Both get as well as the cast can result in a nullpointer. In both cases we
+		// return false.
+		MainPanel* mainPanel = dynamic_cast<MainPanel*> (gamePanels.Root().get());
+		if(!mainPanel)
+			return false;
+		
+		mainPanel->GiveCommand(command);
+		return true;
+	}
+}
+
+
+
+TestStep::TestStep()
+{
+	// Initialize with some sensible default values.
+	stepType = 0;
+	filePathOrName = "";
+}
+
+
+
+TestStep::~TestStep()
+{
+}
+
+
+
+int TestStep::StepType()
+{
+	return stepType;
+}
+
+
+
+const string TestStep::FilePathOrName()
+{
+	return filePathOrName;
+}
+
+
+
+bool TestStep::PlayerIsFlyingAround(UI &menuPanels, UI &gamePanels, PlayerInfo &player)
+{
+	// Code borrowed from main.cpp
+	bool inFlight = (menuPanels.IsEmpty() && gamePanels.Root() == gamePanels.Top());
+	return inFlight;
+}
+
+
+
+
+
+
+
+bool TestStep::PlayerMenuIsActive(UI &menuPanels)
+{
+	return !menuPanels.IsEmpty();
+}
+
+
+
+bool TestStep::PlayerOnPlanetMainScreen(UI &menuPanels, UI &gamePanels, PlayerInfo &player)
+{
+	if (! menuPanels.IsEmpty())
+		return false;
+	if (gamePanels.Root() == gamePanels.Top())
+		return false;
+
+	PlanetPanel *topPlanetPanel = GetPlanetPanelIfAvailable(gamePanels);
+	return topPlanetPanel != nullptr;
+}
+
+
+
+PlanetPanel * TestStep::GetPlanetPanelIfAvailable(UI &gamePanels)
+{
+	if (gamePanels.IsEmpty()){
+		return nullptr;
+	}
+	Panel *topPanel = gamePanels.Top().get();
+	// If we can cast the topPanel from the gamePanels to planetpanel,
+	// then we have landed and are on a planet.
+	return dynamic_cast<PlanetPanel*>(topPanel);
+}
+
+
+
+void TestStep::Load(const DataNode &node)
+{
+	if (node.Token(0) == "load" and node.Size() > 1)
+	{
+		stepType = LOAD_GAME;
+		filePathOrName = node.Token(1);
+	}
+	else if (node.Token(0) == "assert")
```

Comment:
Reorder the branches to operate with specific logic: first keys that require no second token, then those should have a second, then those that should have a third, etc. 
In this manner you can add a branch between the types to catch all uses that should have something but don't, and avoid needing to continually check the node size

Reply:
Good idea. I think I have seen this done in other parts of the code as well.